### PR TITLE
bmm dram sharded opt

### PR DESCRIPTION
**Pull request by yugaoTT**

### CODE REVIEW

File: tt_metal/common/metal_soc_descriptor.cpp [hunk 1]
```diff
@@ -109,10 +109,22 @@ CoreCoord metal_SocDescriptor::get_physical_tensix_core_from_logical(const CoreC
...
 
+CoreCoord metal_SocDescriptor::get_physical_dram_core_from_logical(const CoreCoord &logical_coord) const {
+    TT_FATAL(
+        (logical_coord.x < this->get_num_dram_channels()) and
+        (logical_coord.y == 0),
+        "Bounds-Error -- Logical_core={} is outside of logical_grid_size={}",
+        logical_coord.str(),
+        CoreCoord(1, this->get_num_dram_channels())
+    );
+    return this->get_preferred_worker_core_for_dram_channel(logical_coord.x);
```
Comment (bbradelTT) [inline tt_metal/common/metal_soc_descriptor.cpp:120]: I remember a discussion about GS vs WH dram locations being vertical vs horizontal.

Why only use x and not y, based on the architecture, and what documentation would support that for each architecture (GS, WH, BH)?

Comment (yugaoTT) [reply tt_metal/common/metal_soc_descriptor.cpp:120]: hey, this has been merged to main, but the reason is that we think of the logical dram grid as 1d. and this helps a lot when using the current sharding infra

**Review by arakhmati — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+    in0 = torch.randn(in0_shape).bfloat16().float()
+    # step = K // num_cores
+    # in0 = torch.ones(in0_shape).bfloat16().float()
+    # for i in range(num_cores):  # since 32768 / 16 = 2048
+    #     in0[:, :, :, i * step : (i + 1) * step] = i + 1
+    in1 = torch.randn(in1_shape).bfloat16().float()
+    # in1 = torch.ones(in1_shape).bfloat16().float()
+    bias = torch.randn(bias_shape).bfloat16().float()
+    # bias = torch.ones(bias_shape).bfloat16().float() * 10
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py:162]: Clean this up?

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+    pt_out = in0 @ in1
+    if has_bias:
+        pt_out += bias
+
+    tt_out = tt2torch_tensor(output_t)
+
+    print(pt_out)
+    print(tt_out)
+
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py:248]: Remove

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        (False, True, True, 32, 8192, 1280, None, (8, 1)),
+        (False, True, True, 32, 8192, 4096, None, (8, 2)),
+        (False, True, True, 32, 8192, 1024, None, (8, 1)),
+        (False, True, True, 32, 32768, 1024, None, (8, 2)),
+        # (False, True, True, 32, 4096, 6144, None, (8, 2), ttl.tensor.DataType.BFLOAT16, ttl.tensor.DataType.BFLOAT8_B, ttl.tensor.DataType.BFLOAT16),
+        # (False, True, True, 32, 4096, 14336, None, (8, 2), ttl.tensor.DataType.BFLOAT16, ttl.tensor.DataType.BFLOAT4_B, ttl.tensor.DataType.BFLOAT8_B),
+        # (False, True, True, 32, 14336, 4096, None, (8, 2), ttl.tensor.DataType.BFLOAT8_B, ttl.tensor.DataType.BFLOAT8_B, ttl.tensor.DataType.BFLOAT8_B),
+        # (False, True, True, 32, 4096, 14336, None, (8, 2), ttl.tensor.DataType.BFLOAT16, ttl.tensor.DataType.BFLOAT4_B, ttl.tensor.DataType.BFLOAT8_B),
+    ],
+)
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py:85]: Clean this up

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp [hunk 5]
```diff
@@ -1013,6 +1014,37 @@ void Matmul::validate(
...
+                uint32_t per_core_M = program_config.per_core_M;
+                uint32_t per_core_N = program_config.per_core_N;
+                auto shard_shape = input_tensor_a.shard_spec().value().shape;
+
+                // No padding
+                TT_FATAL(M == per_core_M);
+                TT_FATAL(per_core_M == (shard_shape[0] / TILE_HEIGHT));
+                TT_FATAL(K % program_config.in0_block_w == 0);
+                TT_FATAL((shard_shape[1] / TILE_WIDTH) % program_config.in0_block_w == 0);
+                // TT_FATAL(div_up(N, per_core_N) == input_tensor_a.shard_spec().value().grid.num_cores());
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:1042]: remove

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp [hunk 6]
```diff
@@ -1197,6 +1245,32 @@ std::vector<Tensor> Matmul::create_output_tensors(const std::vector<Tensor>& inp
...
+                                         MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfig>) {
+                    uint32_t M =
+                        (program_config.fuse_batch ? input_tensor_a.volume() / input_tensor_a.get_legacy_shape()[-1]
+                                                   : input_tensor_a.get_legacy_shape()[-2]) /
+                        TILE_HEIGHT;
+                    uint32_t N = input_tensor_b.get_legacy_shape()[-1] / TILE_WIDTH;
+                    auto input_tensor_b_shape = input_tensor_b.get_legacy_shape();
+                    uint32_t N_unpaddded = input_tensor_b.get_legacy_shape()[-1] -
+                                           input_tensor_b_shape.padding()[input_tensor_b_shape.rank() - 1].back;
+
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:1259]: Why does the matmul need to be aware of padded inputs? If input is padded, output should also be padded.

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp [hunk 7]
```diff
@@ -1424,6 +1517,43 @@ Tensor matmul_1d(
...
+                mem_config,
+                output_dtype,
+                kernel_config_val,
+                untilize_out)};
+        },
+        {input_tensor_a, input_tensor_b},
+        output_tensors,
+        {bias});
+    return output_tensors.at(0);
+}
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:1555]: Can we put this into the primary matmul path?It doesn't seem like it's doing anything new here.

File: tt_eager/tt_dnn/op_library/bmm/kernels/compute/bmm_large_block_zm_fused_bias_activation.cpp [hunk 8]
```diff
@@ -15,6 +15,8 @@
 
 #include "compute_kernel_api/eltwise_unary/sfpu_split_includes.h"
 
+// #include "debug/dprint.h"
+
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/kernels/compute/bmm_large_block_zm_fused_bias_activation.cpp:20]: Remove

File: tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+#include "dataflow_api.h"
+#include "hostdevcommon/common_values.hpp"
+
+// #include "debug/dprint.h"
+
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp:10]: Remove

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.hpp [hunk 10]
```diff
@@ -215,6 +216,44 @@ struct MatmulMultiCoreReuseMultiCast1DProgramConfig {
...
+    std::size_t in0_block_w;
+    std::size_t out_subblock_h;
+    std::size_t out_subblock_w;
+    std::size_t per_core_M;
+    std::size_t per_core_N;
+    bool fuse_batch;
+    std::optional<UnaryWithParam> fused_activation;
+    bool skip_compute;
+    bool skip_in0_mcast;
+    bool skip_write_back;
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.hpp:229]: What are these knobs for? Are they needed?

File: tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        in0_mcast_dest_noc_start_x,
+        in0_mcast_dest_noc_start_y,
+        in0_mcast_dest_noc_end_x,
+        in0_mcast_dest_noc_end_y,
+        0);
+
+    uint32_t local_read_addr = get_read_ptr(cb_id_in2);
+
+    if (not is_worker_core) {
+
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp:76]: We should `constexpr` all paths like this

File: tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in1_sender_dram_sharded.cpp [hunk 12]
```diff
@@ -0,0 +1,170 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+#include "dataflow_api.h"
+#include "hostdevcommon/common_values.hpp"
+
+// #include "debug/dprint.h"
+
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in1_sender_dram_sharded.cpp:10]: Remove

**Review by TT-BrianLiu — CHANGES_REQUESTED** [review]
Couple of things that I think need addressing:
- We should not add another matmul op. You can directly call all primary matmul (or micro-ops) through the shared matmul path now.
- I don't think any matmul should output tensor with shapes that strip away padding under the hood. If we want padding feature, we should make it internal to the matmul or just have output be padded.
- Can we review what the three `skip_compute`, `skip_in0_mcast`, and `skip_write_back` flags do? Are they for testing or actually control critical functionality of the matmul?
- It looks like there's a lot of new kernels with very similar code. Are we sure that they all need to be forked and duplicated?

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp (hunk 5 above)
Comment (yugaoTT) [reply tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:1042]: why remove the checks? is it checked somewhere else?

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp (hunk 6 above)
Comment (yugaoTT) [reply tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:1259]: the in1 reader needs it, so it doesn't send back un-used data to storage cores

**Comment by yugaoTT** [comment]
> Couple of things that I think need addressing:
> 
> * We should not add another matmul op. You can directly call all primary matmul (or micro-ops) through the shared matmul path now.
> * I don't think any matmul should output tensor with shapes that strip away padding under the hood. If we want padding feature, we should make it internal to the matmul or just have output be padded.
> * Can we review what the three `skip_compute`, `skip_in0_mcast`, and `skip_write_back` flags do? Are they for testing or actually control critical functionality of the matmul?
> * It looks like there's a lot of new kernels with very similar code. Are we sure that they all need to be forked and duplicated?

Hey,
for 1, I'll change it to call matmul directly.
for 2, the strip away is not needed, it is old code for testing, I'll remove that.
for 3, these knobs are useful if we want to benchmarking for blackhole, and easier for other people to test.
for 4, I think it is better to stay separated code, in case we need debugging later.

**Review by eyonland — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py (hunk 4 above)
Comment (yugaoTT) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py:85]: will leave it here for now, in case more benchmarking needed

File: tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp (hunk 11 above)
Comment (yugaoTT) [reply tt_eager/tt_dnn/op_library/bmm/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp:75]: it is run-time arg, can't use

**Review by TT-BrianLiu — APPROVED** [review]
Once the internal padding works, it should be good to merge.

**Review by bbradelTT — APPROVED** [review]
